akinet/tls/client_parser.go

```diff
@@ -0,0 +1,273 @@
+
+func (*tlsClientHelloParser) Name() string {
+	return "TLS 1.2/1.3 Client-Hello Parser"
+}
+
+func (parser *tlsClientHelloParser) Parse(input memview.MemView, isEnd bool) (result akinet.ParsedNetworkContent, unused memview.MemView, err error) {
+	result, numBytesConsumed, err := parser.parse(input, isEnd)
+	// It's an error if we're at the end and we don't yet have a result.
+	if isEnd && result == nil && err == nil {
+		// We never got the full TLS record. This is an error.
+		err = errors.New("incomplete TLS record for Client Hello")
+	}
+
+	// If we have an error, then we cannot consume any input according to the
+	// contract for Parse.
+	if err != nil {
+		numBytesConsumed = 0
+	}
+
+	unused = parser.allInput.SubView(numBytesConsumed, parser.allInput.Len())
+	return result, unused, err
+}
+
+func (parser *tlsClientHelloParser) parse(input memview.MemView, isEnd bool) (result akinet.ParsedNetworkContent, numBytesConsumed int64, err error) {
+	// Add the incoming bytes to our buffer.
+	parser.allInput.Append(input)
+
+	// Wait until we have at least the TLS record header.
+	if parser.allInput.Len() < tlsRecordHeaderLength_bytes {
+		return nil, 0, nil
+	}
+
+	// The last two bytes of the record header give the total length of the
+	// handshake message that appears after the record header.
+	handshakeMsgLen_bytes := parser.allInput.GetUint16(tlsRecordHeaderLength_bytes - 2)
+	handshakeMsgEndPos := int64(tlsRecordHeaderLength_bytes + handshakeMsgLen_bytes)
+
+	// Wait until we have the full handshake record.
+	if parser.allInput.Len() < handshakeMsgEndPos {
+		return nil, 0, nil
+	}
+
+	// Get a Memview of the handshake record and a corresponding reader.
+	buf := parser.allInput.SubView(tlsRecordHeaderLength_bytes, handshakeMsgEndPos)
+	reader := buf.CreateReader()
+
+	// Seek past some headers.
+	_, err = reader.Seek(handshakeHeaderLength_bytes+clientVersionLength_bytes+clientRandomLength_bytes, io.SeekCurrent)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Now at the session ID, which is a variable-length vector. The first byte
+	// indicates the vector's length in bytes.
+	sessionIdLen_bytes, err := reader.ReadByte()
+	if err != nil {
+		return nil, 0, err
+	}
+	_, err = reader.Seek(int64(sessionIdLen_bytes), io.SeekCurrent)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Now at the cipher suites. The first two bytes gives the length of this
+	// header in bytes.
+	cipherSuitesLen_bytes, err := reader.ReadUint16()
+	if err != nil {
+		return nil, 0, err
+	}
+	_, err = reader.Seek(int64(cipherSuitesLen_bytes), io.SeekCurrent)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	// Now at the compression methods. The first byte gives the length of this
+	// header in bytes.
+	compressionMethodsLen_bytes, err := reader.ReadByte()
+	if err != nil {
+		return nil, 0, err
+	}
+	_, err = reader.Seek(int64(compressionMethodsLen_bytes), io.SeekCurrent)
+	if err != nil {
+		return nil, 0, err
+	}
```

Comment:
Probably a helper method would have saved some copy/paste here?

Reply:
In 9155756.